agent/lib/kontena/launchers/etcd.rb

```diff
@@ -1,141 +1,147 @@
     include Kontena::Helpers::IfaceHelper
 
     ETCD_VERSION = ENV['ETCD_VERSION'] || '2.3.7'
     ETCD_IMAGE = ENV['ETCD_IMAGE'] || 'kontena/etcd'
+    IMAGE = "#{ETCD_IMAGE}:#{ETCD_VERSION}"
 
     def initialize(autostart = true)
-      @image_pulled = false
-      @running = false
-      @image_name = "#{ETCD_IMAGE}:#{ETCD_VERSION}"
       info 'initialized'
-      subscribe('network_adapter:start', :on_overlay_start)
       async.start if autostart
     end
 
     def start
-      pull_image(@image_name)
-    end
+      ensure_image(IMAGE)
 
-    # @param [String] topic
-    # @param [Node] node
-    def on_overlay_start(topic, node)
-      retries = 0
-      begin
-        self.start_etcd(node)
-      rescue Docker::Error::ServerError => exc
-        log_error(exc)
-        if retries < 4
-          retries += 1
-          sleep 0.25
-          retry
-        end
-      rescue => exc
-        log_error(exc)
+      observe(Actor[:node_info_worker], Actor[:weave_launcher]) do |node, weave|
+        # TODO: only update after both node-info and weave-launcher have updated?
+        # XXX: exclusive updates...
+        self.update(node)
       end
     end
 
-    # @param [Node] node
-    def start_etcd(node)
-      sleep 1 until image_pulled?
+    # @param node [Node]
+    def update(node)
+      state = self.ensure(node)
 
-      create_data_container(@image_name)
-      create_container(@image_name, node)
-    end
+      update_observable(state)
 
-    # @param [String] image
-    def pull_image(image)
-      if Docker::Image.exist?(image)
-        @image_pulled = true
-        return
-      end
-      Docker::Image.create('fromImage' => image)
-      sleep 1 until Docker::Image.exist?(image)
-      @image_pulled = true
-    end
+    rescue => exc
+      error exc
 
-    # @return [Boolean]
-    def image_pulled?
-      @image_pulled == true
+      reset_observable
     end
 
-    def running?
-      @running == true
+    # @param [Node] node
+    # @return [Hash{running: true}]
+    def ensure(node)
+      ensure_data_container(IMAGE)
+      container = ensure_container(IMAGE, node)
+
+      {
+        container_id: container.id,
+        overlay_ip: node.overlay_ip,
+        dns_name: 'etcd.kontena.local',
+      }
     end
 
     # @param [String] image
-    def create_data_container(image)
-      data_container = Docker::Container.get('kontena-etcd-data') rescue nil
-      unless data_container
-        Docker::Container.create(
+    # @return [Docker::Container]
+    def ensure_data_container(image)
+      unless data_container = inspect_container('kontena-etcd-data')
+        info "creating new etcd data container"
+
+        data_container = Docker::Container.create(
           'name' => 'kontena-etcd-data',
           'Image' => image,
           'Volumes' => {'/var/lib/etcd' => {}}
         )
       end
     end
 
+    # TODO: inspect --listen-client-urls and ensure node.overlay_ip, docker_ip matches (#1893)
```

Comment:
#1893 the `kontena-etcd` container should be re-created if the `docker0` or `weave` bridge IPs change